doc/source/serve/doc_code/streaming_tutorial.py

```diff
@@ -0,0 +1,348 @@
+# flake8: noqa
+# fmt: off
+
+from typing import List
+
+# __textbot_setup_start__
+import asyncio
+import logging
+from queue import Empty
+
+from fastapi import FastAPI
+from starlette.responses import StreamingResponse
+from transformers import AutoModelForCausalLM, AutoTokenizer, TextIteratorStreamer
+
+from ray import serve
+
+logger = logging.getLogger("ray.serve")
+# __textbot_setup_end__
+
+
+# __textbot_constructor_start__
+fastapi_app = FastAPI()
+
+
+@serve.deployment
+@serve.ingress(fastapi_app)
+class Textbot:
+    def __init__(self, model_id):
+        self.loop = asyncio.get_running_loop()
+
+        self.model_id = model_id
+        self.model = AutoModelForCausalLM.from_pretrained(self.model_id)
+        self.tokenizer = AutoTokenizer.from_pretrained(self.model_id)
+
+    # __textbot_constructor_end__
+
+    # __textbot_logic_start__
+    @fastapi_app.post("/")
+    def handle_request(self, prompt: str):
+        logger.info(f'Got prompt: "{prompt}"')
+        streamer = TextIteratorStreamer(
+            self.tokenizer, timeout=0, skip_prompt=True, skip_special_tokens=True
+        )
+        self.loop.run_in_executor(None, self.generate_text, prompt, streamer)
+        return StreamingResponse(
+            self.consume_streamer(streamer), media_type="text/plain"
+        )
+
+    def generate_text(self, prompt: str, streamer: TextIteratorStreamer):
+        input_ids = self.tokenizer([prompt], return_tensors="pt").input_ids
+        self.model.generate(input_ids, streamer=streamer, max_length=10000)
+
+    async def consume_streamer(self, streamer: TextIteratorStreamer):
```

Comment:
this is ugly D:

Reply:
Yeah– ideally `streamer` would be an async iterator, but I think this is necessary without that.